Add polkit 🔒 support

kalev opened the pull request:
Here's my very first try on contributing to rpm-ostree; please let me know if I should do anything differently.

This pull request adds polkit support to org.projectatomic.rpmostree1.OS interface which so far has been completely locked down to root-only. We need this for landing the gnome-software rpm-ostree backend where gnome-software runs as a regular user.

Not sure how granular the individual polkit actions should be. I tried to keep them roughly matching in granularity with what PackageKit and flatpak are already doing, except for one exception: package install and uninstall.

I lumped install and uninstall together under one action here because the rpm-ostree DBus API allows doing them in one transaction. The main reason why they were separate for PackageKit and flatpak was to allow wheel group users to install apps without popping up a polkit dialog, but at the same time we wanted to require auth for uninstall. I think since we're going for an app / base OS split in rpm-ostree based Fedora Workstation, it should be fine to support auth-less app install only with flatpak and require auth for the rpm-ostree layer app install, but if needed we should be able to easily change this in the future.

Comments welcome!

cgwalters commented:
bot, add author to whitelist

jlebon commented on src/daemon/org.projectatomic.rpmostree1.policy:
```diff
@@ -0,0 +1,87 @@
+    <defaults>
+      <allow_any>auth_admin</allow_any>
+      <allow_inactive>auth_admin</allow_inactive>
+      <allow_active>yes</allow_active>
```
I understand why this is `yes`, but now anyone can upgrade, right? My polkit-foo is not very strong, but is there not a way to only allow gnome-software to upgrade? (Since I assume gnome-software already does its own checks for authorized users, right?).

cgwalters replied:
Not quite - the user has to be on a *physical console*.  This is a crucial distinction that was settled on long, long in the past for "can an unprivileged user reboot the machine" for example.

Specifically compare to `org.freedesktop.login1.reboot`.

[There's no sane way to distinguish between calling applications](https://bugzilla.gnome.org/show_bug.cgi?id=533493).

cgwalters replied:
Of course since I filed that bugzilla, we are working on flatpak, which eventually will provide exactly this type of strong identity for applications, although right now the flatpak security policy is still full of holes (if one allows access to the homedir etc.)

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +81,101 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+  else if (g_strcmp0 (method_name, "Deploy") == 0 ||
+           g_strcmp0 (method_name, "Rebase") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.rebase";
```
Hmm, I'm thinking now it'd make more sense to put `Deploy` together with `Upgrade`. Both of them are constrained to stay within the same branch (after https://github.com/projectatomic/rpm-ostree/pull/495), whereas `Rebase` is much more powerful.

`Deploy` can also be used as a safer race-free version of `Upgrade`, as used by Cockpit (and probably gnome-software should use `Deploy` as well here if e.g. pending versions/pkg diff previews etc... are displayed).

cgwalters replied:
Offhand I'd say in the "organization owns desktop" case I'd want a separation between `Deploy` and `Upgrade`.  Really in that case I think a lot of organizations would want automatic update downloads and only allow the user to delay the apply step.

(Supporting that is its own whole topic)

cgwalters commented:
The CI failing is https://bugzilla.redhat.com/show_bug.cgi?id=1460094

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +81,101 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+    {
+      action = "org.projectatomic.rpmostree1.package-install-uninstall";
+    }
+  else if (g_strcmp0 (method_name, "UpdateDeployment") == 0)
```
Yeah, this is going to make this especially tricky now that every method can end up using `UpdateDeployment`!

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +81,101 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+          g_strv_length (uninstall_pkgs) > 0)
+        action = "org.projectatomic.rpmostree1.package-install-uninstall";
+      else
+        action = "org.projectatomic.rpmostree1.upgrade";
```
This will require at least another `else if` to check if it's a `rebase` action instead. E.g. by checking if the `set-refspec` modifier is used.

cgwalters replied:
This part is also going to be a pain to maintain =(  For similar reasons as https://github.com/projectatomic/rpm-ostree/pull/814

Perhaps we can extract a helper function to parse `UpdateDeployment` invocation arguments and use it in both places?

cgwalters commented on src/daemon/rpmostreed-os.c:
```diff
@@ -104,9 +204,13 @@ static void
+  self->authority = polkit_authority_get_sync (NULL, &error);
+  if (self->authority == NULL)
+    {
+      g_error ("Can't get polkit authority: %s", error->message);
```
I tend to avoid `g_error()` in general because few conditions where one typically finds it necessitate dumping core.  That's true here; if we fail to talk to polkitd, it's highly unlikely that a core dump from rpm-ostreed is going to be useful to operators.

Hence I'd prefer `errx()` in this case.

(And in cases where we *do* want a core dump, that's `g_assert()`)

kalev commented:
Different from other 3 CI failures, c7-build failed with missing polkit-gobject-1 on the build server. Any idea how I could get it to install it? I did add "BuildRequires: pkgconfig(polkit-gobject-1)" to the spec.in file in git, but looks like it's not used for installing build deps?

cgwalters commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +81,101 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+      g_strcmp0 (method_name, "GetCachedUpdateRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "DownloadUpdateRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "GetCachedRebaseRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "DownloadRebaseRpmDiff") == 0)
```
This looks OK, but I wonder if there's a best practice for embedding the "action binding" with the method itself?  Maybe something like including it in an annotation in the XML and having some way to get at that?  I think I saw that systemd defines structs for its methods, but I'm not sure how easy that is to do with GDBus.

kalev replied:
My goal here was to have a less locked down function that gnome-software could use to prepare an update in background, and then only pop up a polkit dialog when everything is ready to go and the user clicks update. Similar to how the offline updates currently work -- gnome-software downloads rpms in the background, prepares everything and then the user has to click 'Restart & Install' to actually do the update (which is a natural point to pop up a polkit dialog as well).

jlebon commented:
Heh yeah, the spec file there is not really canonical. You can add it directly in the `.papr.yml` file, in the `c7-build` section, where we do a `yum install -y make`.

jlebon commented:
bot, retest this please

cgwalters commented:
Perhaps we should discuss this in https://github.com/projectatomic/rpm-ostree/issues/745 - but I'd like to spend some time thinking about the polkit defaults.  Defaults are always highly influential.

One major topic here is that for many, many years in Fedora, there's been a distinction between `yum/dnf` and `PackageKit`, where only the latter speaks polkit.  While Fedora Server *does* make `PackageKit` mandatory, I know there are quite a lot of server people who start from "minimal" installations and hence don't have `PK`.  And if you don't have `PackageKit` installed, then as a server operator it's all really simple to think about - it devolves to the "basic timesharing Unix model" of changing the OS requiring root.

Obviously for us, Atomic Host is intended to be minimal (and also has not included `PackageKit`).  The proposed polkit policy here includes several `allow_active=yes`.  (I see Jonathan commented on this too).

I'm not opposed exactly but I'd like to make sure we consider this carefully - I wonder how much it makes sense to share defaults between server and workstation cases.  (A long-running topic in the workstation world is the separation between "user owns machine" and "organization owns machine"; in the latter one would want to turn off package installs/removals commonly for example)

kalev commented:
Thanks for the comments! Sure, perhaps it makes sense to have different defaults for server and workstation -- I'll reply in the issue ticket to keep the discussion in one place.

kalev commented:
OK, pushed a fixup that I think addresses all the comments so far:

1) changes all `yes` actions to `auth_admin_keep`
2) uses `errx()` instead of `g_error()`
3) moves `Deploy` so that it's handled the same as `Upgrade`
4) checks `set-refspec` in the `UpdateDeployment` handler and assumes we're rebasing then
5) adds a new `install-local-packages` action
6) renames `package-install-uninstall` to `install-uninstall-packages` to match with the newly added `install-local-packages` action

jlebon commented on .papr.yml:
```diff
@@ -41,6 +41,7 @@ extra-repos:
 # XXX: integrate better with ci/ scripts
 tests:
   - yum-builddep -y rpm-ostree
+  - yum install -y polkit-devel
```
Ahh right, we'll also need this at least temporarily added to `install_builddeps` in `libbuild.sh`. Hopefully we can eventually make the spec file in this repo be the canonical one to make this easier.

jlebon commented on src/daemon/org.projectatomic.rpmostree1.policy:
```diff
@@ -0,0 +1,98 @@
+    </defaults>
+  </action>
+
+  <action id="org.projectatomic.rpmostree1.install-local-packages">
```
Hmm, I think it does make sense to make this separate from `install-uninstall-packages`, since in the latter case the package has to come from a repo in `/etc/yum.repos.d`. Though note that right now at least, uninstalling local packages falls under the same perms as `install-uninstall-packages`.

kalev replied:
The reason why I split local package installing out was that I think it's inherently a more unsafe operation than installing from a repo. For instance, I think it might make sense to make installing from a repo work without password prompt in Fedora Workstation, but when someone wants to install a random package downloaded from the internet (and not signed with our gpg keys etc), we should ask for confirmation.

I think it's fine that there's only one uninstall action, I don't think there's a need for separate uninstall-local-packages

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +82,109 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+      action = "org.projectatomic.rpmostree1.rebase";
+    }
+  else if (g_strcmp0 (method_name, "Deploy") == 0 ||
+           g_strcmp0 (method_name, "Upgrade") == 0)
```
I think @cgwalters was arguing for making these two separate actions.

kalev replied:
OK, let me make that so.

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +82,109 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+      else if (install_local_pkgs != NULL)
+        action = "org.projectatomic.rpmostree1.install-local-packages";
+      else
+        action = "org.projectatomic.rpmostree1.upgrade";
```
Hmm, yeah these checks are messy indeed since multiple actions can happen. Is it possible here to accumulate actions and then check that the user is authorized for all of them? Otherwise it's just going to be messy how to determine the "most privileged action", which itself might depend on e.g. the site policy.

kalev replied:
Sure, I think I could just ask polkit for several permissions in a row. That's a good idea, let me do that. I don't think we can lock down a do-everything method like UpdateDeployment without that.

Although when considering user experience, having several password prompts in a row is probably not going to look very nice. We should probably somehow make sure that in normal use case when a regular user is using rpm-ostree cli we don't pop up several password prompts up in a row.

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +85,131 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+        g_ptr_array_add (actions, "org.projectatomic.rpmostree1.install-uninstall-packages");
+
+      if (install_local_pkgs != NULL && g_variant_n_children (install_local_pkgs) > 0)
+        g_ptr_array_add (actions, "org.projectatomic.rpmostree1.install-local-packages");
```
Hmm, we'll need another check here for the override related modifiers, under e.g. a new `org.projectatomic.rpmostree1.override` action? Not sure if it's worth splitting into multiple actions just yet. They're experimental, but don't use the experimental interface like livefs.

kalev replied:
Sure, let me do that, thanks

kalev replied:
Done; let's see if the tests pass now

jlebon commented:
Awesome, thanks for the patch!
@rh-atomic-bot r+ c49b58b

cgwalters commented:
Looks like this is likely to bounce:

> error: rpmostreed OS operation UpdateDeployment not allowed for user

Can we give this a day to merge?  I'd like a chance to review carefully too but I'm in meetings all day, I might be able to look later or tonight.

cgwalters commented:
(Not opposed to landing in master and doing fixups after if @jlebon is happy but hopefully a day is OK)

jlebon commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +85,144 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+
+      if (override_replace_pkgs != NULL || override_remove_pkgs != NULL || override_reset_pkgs != NULL ||
+          (override_replace_local_pkgs != NULL && g_variant_n_children (override_replace_local_pkgs) > 0))
+        g_ptr_array_add (actions, "org.projectatomic.rpmostree1.override");
```
Oh right, this also needs to check for the `no-overrides` flag.

kalev replied:
I've got this fixed up locally, just waiting to see what happens with the ongoing merge to see if I should put it here or create a new PR.

jlebon commented:
Sure, that's good with me.

kalev commented:
I'm fine either way, thanks for the reviews!

cgwalters commented on src/daemon/org.projectatomic.rpmostree1.policy:
```diff
@@ -0,0 +1,120 @@
+    </defaults>
+  </action>
+
+  <action id="org.projectatomic.rpmostree1.set-initramfs-state">
```
Maybe we could have this be a generic `bootconfig` permission, in preparation for https://github.com/projectatomic/rpm-ostree/issues/594

kalev replied:
Sure, let me change that. `bootconfig` feels like a much better name to me, thanks!

cgwalters commented on src/daemon/rpmostreed-os.c:
```diff
@@ -77,6 +85,147 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+          (override_replace_local_pkgs != NULL && g_variant_n_children (override_replace_local_pkgs) > 0) ||
+          no_overrides)
+        g_ptr_array_add (actions, "org.projectatomic.rpmostree1.override");
+    }
```
I'd feel better if we had an:
```
else
 {
   authorized = FALSE;
 }
```
here.

kalev replied:
Sure, easy enough to add, although we already initialize it to `FALSE` in the `gboolean authorized = FALSE;` line above, which makes having the `else` block unnecessary.

cgwalters replied:
Yep, I did see that, but it's more "obviously correct" to have the `else`.

kalev commented:
All fixed up!

cgwalters commented:
Thanks a lot for this!

@rh-atomic-bot r+ 26ed454

kalev commented:
Thanks!
